### ISSUE TITLE
Check that lab works on Python 3.8

### ISSUE THREAD

**Issue by vidartf** [jupyterlab/jupyterlab#6487]
# Summary

Works in Python 3.8 on:

- [ ] Windows
- [ ] Linux
- [X] Mac
---
# Original description

Python 3.8 release is right around the corner, so we should do a quick check that nothing obviously breaks on either of the three major platforms (*nix, OSX, Windows).

**Comment by stonebig** [comment]
1rst try on jupyterlab-1.0.0rc0 + Notebook-6.0.rc1 + Tornado-6.0.3 on Windows:
- I need to replace one line in site-packages\tornado\platform\asyncio.py:
  .  `import asyncio`
  . `import asyncio;asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())  `
- then jupyterlab is stuck at this, after a few seconds of Jupyter planets movements
![image](https://user-images.githubusercontent.com/4312421/60327164-7db49000-998b-11e9-8a9a-e309e7ce36cb.png)

**Comment by stonebig** [comment]
maybe I must get ipywidgets-7.5.0rc +  widgetsnbextension-3.5.0rc0 ?

**Comment by stonebig** [comment]
still failing while using the latest jupyterlab/commands.py.... I suppose that part is not a python-3.8 problem

**Comment by vidartf** [comment]
What happens if you try in a fresh env without any extensions? Does it work?

**Comment by stonebig** [comment]
not used /not good to extension-less env. you asked last time but I failed to make it work. the problem may have been also elsewhere at that time.

**Comment by stonebig** [comment]
for the python-3.8 part, the only problem seems that ugly patch needed on tornado, that may be to replace per something from a jupyter library. https://bugs.python.org/issue37373

**Comment by stonebig** [comment]
same error with last hour release.... what do I miss ?

````
Package                       Version     
----------------------------- ------------
adodbapi                      2.6.0.7     
alabaster                     0.7.12      
altair                        3.1.0       
altair-widgets                0.2.0       
ansiwrap                      0.8.4       
appdirs                       1.4.3       
asciitree                     0.3.3       
asteval                       0.9.14      
astroid                       2.2.5       
atomicwrites                  1.3.0       
attrs                         19.1.0      
autopep8                      1.4.4       
Babel                         2.7.0       
backcall                      0.1.0       
backports-abc                 0.5         
baresql                       0.7.4       
beautifulsoup4                4.7.1       
black                         19.3b0      
bleach                        3.1.0       
blosc                         1.8.1       
bloscpack                     0.16.0      
bokeh                         1.2.0       
Bottleneck                    1.2.1       
bqplot                        0.11.5      
branca                        0.3.1       
brewer2mpl                    1.4.1       
Brotli                        1.0.7       
certifi                       2019.6.16   
cffi                          1.12.3      
cftime                        1.0.3.4     
chardet                       3.0.4       
Click                         7.0         
cloudpickle                   1.2.1       
clrmagic                      0.0.1a2     
colorama                      0.4.1       
cycler                        0.10.0      
Cython                        3.0a0       
dask                          2.0.0       
dask-searchcv                 0.2.0       
db.py                         0.5.3       
decorator                     4.4.0       
defusedxml                    0.6.0       
Deprecated                    1.2.5       
descartes                     1.1.0       
distributed                   2.0.0       
docopt                        0.6.2       
docrepr                       0.1.1       
docutils                      0.14        
entrypoints                   0.3         
fasteners                     0.15        
Flask                         1.0.3       
future                        0.17.0      
greenlet                      0.4.15      
h5py                          2.9.0       
HeapDict                      1.0.0       
holoviews                     1.12.3      
html5lib                      1.0.1       
hvplot                        0.4.0       
idlex                         1.18        
idna                          2.8         
imageio                       2.5.0       
imagesize                     1.1.0       
imbalanced-learn              0.4.3       
importlib-metadata            0.18        
ipydatawidgets                4.0.0       
ipykernel                     5.1.1       
ipyleaflet                    0.10.8      
ipympl                        0.2.1       
ipyparallel                   6.2.4       
ipython                       7.5.0       
ipython-genutils              0.2.0       
ipython-sql                   0.3.9       
ipywidgets                    7.5.0rc1    
isort                         4.3.21      
itsdangerous                  1.1.0       
jedi                          0.14.0      
Jinja2                        2.10.1      
joblib                        0.13.2      
json5                         0.8.4       
jsonschema                    3.0.1       
julia                         0.4.1       
jupyter                       1.0.0       
jupyter-client                5.2.4       
jupyter-console               6.0.0       
jupyter-core                  4.5.0       
jupyter-sphinx                0.1.4       
jupyterlab                    1.0.0       
jupyterlab-launcher           0.13.1      
jupyterlab-server             1.0.0       
kiwisolver                    1.1.0       
lazy-object-proxy             1.4.0       
lmfit                         0.9.13      
locket                        0.2.0       
lxml                          4.3.4       
Markdown                      3.1.1       
MarkupSafe                    1.1.1       
matplotlib                    3.1.0       
mccabe                        0.6.1       
metakernel                    0.24.2      
mistune                       0.8.4       
mizani                        0.5.4       
mkl-service                   2.0.2       
mlxtend                       0.16.0      
mock                          3.0.5       
monotonic                     1.5         
more-itertools                7.0.0       
mpld3                         0.3         
mpldatacursor                 0.6.2       
mpmath                        1.1.0       
msgpack                       0.6.1       
mysql-connector-python        8.0.11      
nbconvert                     5.5.0       
nbformat                      4.4.0       
netCDF4                       1.5.1.2     
networkx                      2.3         
notebook                      6.0.0rc1    
numcodecs                     0.6.3       
numexpr                       2.6.9       
numpy                         1.16.4+mkl  
numpydoc                      0.9.1       
oct2py                        5.0.4       
octave-kernel                 0.31.0      
packaging                     19.0        
palettable                    3.1.1       
pandas                        0.24.2      
pandocfilters                 1.4.2       
panel                         0.6.0       
papermill                     1.0.1       
param                         1.9.1       
parambokeh                    0.2.3       
paramnb                       2.0.4       
parso                         0.5.0       
partd                         0.3.10      
passlib                       1.7.1       
patsy                         0.5.1       
pdvega                        0.2.1.dev0  
pep8                          1.7.1       
pexpect                       4.6.0+dummy 
pg8000                        1.13.1      
pickleshare                   0.7.5       
Pillow                        6.0.0       
pip                           19.1.1      
pkginfo                       1.5.0.1     
plotnine                      0.5.2.dev1  
pluggy                        0.12.0      
ppci                          0.5.6       
prettytable                   0.7.2       
prometheus-client             0.7.1       
prompt-toolkit                2.0.9       
protobuf                      3.8.0       
psutil                        5.6.3       
ptpython                      2.0.4       
PuLP                          1.6.10      
py                            1.8.0       
PyAudio                       0.2.11      
pybars3                       0.9.6       
pybind11                      2.3.0       
pycodestyle                   2.5.0       
pycosat                       0.6.3       
pycparser                     2.19        
pyct                          0.4.6       
pyflakes                      2.1.1       
pygame                        1.9.6       
Pygments                      2.4.2       
pylint                        2.3.1       
PyMeta3                       0.5.1       
pyodbc                        4.0.26      
PyOpenGL                      3.1.3b2     
pyparsing                     2.4.0       
pyrsistent                    0.15.1      
pyserial                      3.4         
pystache                      0.5.4       
pytest                        4.6.3       
python-dateutil               2.8.0       
python-snappy                 0.5.4       
pythonnet                     2.4.1.dev0  
pythreejs                     2.1.0       
pytz                          2019.1      
pyviz-comms                   0.7.2       
PyWavelets                    1.0.3       
pywinpty                      0.5.5       
PyYAML                        5.1.1       
pyzmq                         18.0.1      
pyzo                          4.7.4       
qtconsole                     4.5.1       
QtPy                          1.8.0       
readme-renderer               24.0        
regex                         2019.6.8    
reportlab                     3.5.23      
requests                      2.22.0      
requests-toolbelt             0.9.1       
rise                          5.5.0       
Rtree                         0.8.3       
ruamel.yaml                   0.15.97     
scikit-image                  0.15.0      
scikit-learn                  0.21.2      
scilab2py                     0.6.2       
scipy                         1.3.0       
seaborn                       0.9.1.dev0  
Send2Trash                    1.5.0       
setuptools                    41.0.1      
simplegeneric                 0.8.1       
simpy                         3.0.11      
six                           1.12.0      
snowballstemmer               1.2.1       
sortedcontainers              2.1.0       
sounddevice                   0.3.13      
soupsieve                     1.9.2       
Sphinx                        2.1.2       
sphinx-rtd-theme              0.4.3       
sphinxcontrib-applehelp       1.0.1       
sphinxcontrib-devhelp         1.0.1       
sphinxcontrib-htmlhelp        1.0.2       
sphinxcontrib-jsmath          1.0.1       
sphinxcontrib-qthelp          1.0.2       
sphinxcontrib-serializinghtml 1.1.3       
SQLAlchemy                    1.3.5       
sqlite-bro                    0.9.1       
sqlparse                      0.3.0       
statsmodels                   0.10.0      
supersmoother                 0.4         
sympy                         1.4         
tables                        3.5.2       
tblib                         1.4.0       
tenacity                      5.0.4       
terminado                     0.8.2       
testpath                      0.4.2       
textwrap3                     0.9.2       
toml                          0.10.0      
toolz                         0.9.0       
tornado                       6.0.3       
tqdm                          4.32.2      
traitlets                     4.3.2       
traittypes                    0.2.1       
twine                         1.13.0      
twitter                       1.17.1      
typed-ast                     1.4.0       
tzlocal                       1.5.1       
uncertainties                 3.1.1       
urllib3                       1.25.3      
vega                          2.3.2       
vega-datasets                 0.7.0       
ViTables                      3.0.0       
wcwidth                       0.1.7       
webencodings                  0.5.1       
Werkzeug                      0.15.4      
wheel                         0.33.4      
widgetsnbextension            3.5.0rc1    
winpython                     2.0.20190519
wrapt                         1.11.2      
xarray                        0.12.1      
xlrd                          1.2.0       
XlsxWriter                    1.1.7       
zarr                          2.3.2       
zict                          1.0.0       
zipp                          0.5.1  
````

````
app dir: C:\WinP\bd38\bu\WPy64-3800b1b\python-3.8.0b1.amd64\share\jupyter\lab
        @jupyter-widgets/jupyterlab-manager v1.0.0-rc.1 enabled  ok
        @pyviz/jupyterlab_pyviz v0.8.0-alpha.1 enabled  ok
        bqplot v0.4.5 enabled  ok
        jupyter-leaflet v0.10.7 enabled  ok
        jupyter-matplotlib v0.3.0 enabled  ok
        jupyter-threejs v2.1.0 enabled  ok
        jupyterlab-datawidgets v6.1.0 enabled  ok
        jupyterlab_bokeh v0.2.0 enabled  ok
​````

**Comment by stonebig** [comment]
ok, so it works on Python-3.8 if I build with zero extension

**Comment by stonebig** [comment]
it's ok now in Python-3.8 with 
![image](https://user-images.githubusercontent.com/4312421/60687315-1aa48b00-9eae-11e9-8a73-02118076cf27.png)

**Comment by stonebig** [comment]
Jupyterlab-2.0 is far away. A simple but official hack would be nice in the interlude. As far as we know, there is zero chance Proactor Loop to be improved before Python-3.9 now. 

https://bugs.python.org/issue37373

**Comment by vidartf** [comment]
@stonebig I'm not sure what the remaining issues here are (I'm struggling to follow the previous posts). As far as I can read it, there is an issue in tornado, and an issue in one or more 3rd party extensions, but none in jupyterlab itself?

**Comment by stonebig** [comment]
Jupyterlab and friends can't work by default on Win10 + Python-3.8, because:
- the async Loop by default has been changed and the new default is not compatible with Tornado at the moment,
- and Tornado won't do anything about it, because not their problem.

So:
-  I do my patch on Tornado, as the only way  know to fix it,
- but, Jupyterlab and friends need:
 - either to do a more complex change in their code,
 - either to fork Tornado.

**Comment by stonebig** [comment]
hum, I didn't notice 1.2 was out,  re-checking...

**Comment by stonebig** [comment]
as of now Jupyterlab-1.2.0:
- it starts, so this problem seems solved,
- yet I can't get Jupyterlab rebuild with plugins, so I have to remain on Jupyterlab-1.x.

**Comment by stonebig** [comment]
and I fail to find in which comit of which Jupyter package the asyncio problem may have been fixed. Very frustrating

**Comment by stonebig** [comment]
so rechecking now, it works with Jupyterlab-1.1.4 without patching Tornado.

**Comment by vtrokhymenko** [comment]
maybe anybody can help solve this problem:
![image](https://user-images.githubusercontent.com/8581044/68591444-73f27b80-0499-11ea-80a3-7ba6e9ca9445.png)

so, why i can use python3.8 at jupyter lab (v.1.2.2)

**Comment by jasongrout** [comment]
I just checked with python 3.8 on macOS:

```sh
conda create -n jlabpy38 python=3.8 jupyterlab
```
and it works great, answering the OP.

As for @vtrokhymenko - check to make sure your kernel is actually python 3.8, not just that python 3.8 is available from the command line:

```
import sys
print(sys.version)
```

**Comment by jasongrout** [comment]
Here is the environment that seemed to work well for me:

```
# Name                    Version                   Build  Channel
appnope                   0.1.0                 py38_1000    conda-forge
attrs                     19.3.0                     py_0    conda-forge
backcall                  0.1.0                      py_0    conda-forge
bleach                    3.1.0                      py_0    conda-forge
ca-certificates           2019.9.11            hecc5488_0    conda-forge
certifi                   2019.9.11                py38_0    conda-forge
decorator                 4.4.1                      py_0    conda-forge
defusedxml                0.6.0                      py_0    conda-forge
entrypoints               0.3                   py38_1000    conda-forge
importlib_metadata        0.23                     py38_0    conda-forge
ipykernel                 5.1.3            py38h5ca1d4c_0    conda-forge
ipython                   7.9.0            py38h5ca1d4c_0    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
jedi                      0.15.1                   py38_0    conda-forge
jinja2                    2.10.3                     py_0    conda-forge
json5                     0.8.5                      py_0    conda-forge
jsonschema                3.1.1                    py38_0    conda-forge
jupyter_client            5.3.3                    py38_1    conda-forge
jupyter_core              4.5.0                      py_0    conda-forge
jupyterlab                1.2.2                      py_0    conda-forge
jupyterlab_server         1.0.6                      py_0    conda-forge
libcxx                    9.0.0                h89e68fa_1    conda-forge
libffi                    3.2.1             h6de7cb9_1006    conda-forge
libsodium                 1.0.17               h01d97ff_0    conda-forge
markupsafe                1.1.1            py38h0b31af3_0    conda-forge
mistune                   0.8.4           py38h0b31af3_1000    conda-forge
more-itertools            7.2.0                      py_0    conda-forge
nbconvert                 5.6.1                    py38_0    conda-forge
nbformat                  4.4.0                      py_1    conda-forge
ncurses                   6.1               h0a44026_1002    conda-forge
notebook                  6.0.1                    py38_0    conda-forge
openssl                   1.1.1d               h0b31af3_0    conda-forge
pandoc                    2.7.3                         0    conda-forge
pandocfilters             1.4.2                      py_1    conda-forge
parso                     0.5.1                      py_0    conda-forge
pexpect                   4.7.0                    py38_0    conda-forge
pickleshare               0.7.5                 py38_1000    conda-forge
pip                       19.3.1                   py38_0    conda-forge
prometheus_client         0.7.1                      py_0    conda-forge
prompt_toolkit            2.0.10                     py_0    conda-forge
ptyprocess                0.6.0                   py_1001    conda-forge
pygments                  2.4.2                      py_0    conda-forge
pyrsistent                0.15.5           py38h0b31af3_0    conda-forge
python                    3.8.0                hd366da7_3    conda-forge
python-dateutil           2.8.1                      py_0    conda-forge
pyzmq                     18.1.0           py38h4bf09a9_0    conda-forge
readline                  8.0                  hcfe32e1_0    conda-forge
send2trash                1.5.0                      py_0    conda-forge
setuptools                41.6.0                   py38_1    conda-forge
six                       1.13.0                   py38_0    conda-forge
sqlite                    3.30.1               h93121df_0    conda-forge
terminado                 0.8.2                    py38_0    conda-forge
testpath                  0.4.4                      py_0    conda-forge
tk                        8.6.9             h2573ce8_1003    conda-forge
tornado                   6.0.3            py38h0b31af3_0    conda-forge
traitlets                 4.3.3                    py38_0    conda-forge
wcwidth                   0.1.7                      py_1    conda-forge
webencodings              0.5.1                      py_1    conda-forge
wheel                     0.33.6                   py38_0    conda-forge
xz                        5.2.4             h1de35cc_1001    conda-forge
zeromq                    4.3.2                h6de7cb9_2    conda-forge
zipp                      0.6.0                      py_0    conda-forge
zlib                      1.2.11            h0b31af3_1006    conda-forge
```

**Comment by vtrokhymenko** [comment]
@jasongrout i work from virtualenv

hm, really, from `sys.version` i run not from python3.8

so, i can't understand that, why i can't work with python3.8 when i create virtualenv from python3.8

**Comment by jasongrout** [comment]
@vtrokhymenko - the question is where the kernel is coming from. `jupyter kernelspec list` (executed from the command line in your virtual environment) should give you an idea where Jupyter is finding kernels.

**Comment by bollwyvl** [comment]
Python 3.8 is (presumably) now generally available on azure:
- https://github.com/microsoft/azure-pipelines-image-generation/issues/1317

**Comment by jasongrout** [comment]
@minrk's PR on notebook looks relevant to the windows case: https://github.com/jupyter/notebook/pull/5047

**Comment by vtrokhymenko** [comment]
@jasongrout hi
maybe u can help

so, i create virtualenv with python3.8

when i run in this env `python` i get 3.8
also, when inside run `sys.version` also get 3.8 python


BUT
when i run Jlab i get from sys another python even when run lab from `python -m`

u know how fixed this error?

**Comment by owen-synergy** [comment]
> without considering the re-build issue, Tornado still needs a patch For Jupyterlab-1.2.0 even to start on Python-3.8/Windows 10:
> 
> ```
> [I 13:27:20.167 LabApp] Writing notebook server cookie secret to C:\WinP\bd38\bu\WPy64-3801b1\settings\runtime\notebook_cookie_secret
> Traceback (most recent call last):
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\runpy.py", line 192, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\scripts\jupyter-lab.exe\__main__.py", line 7, in <module>
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\site-packages\jupyter_core\application.py", line 268, in launch_instance
>     return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\site-packages\traitlets\config\application.py", line 663, in launch_instance
>     app.initialize(argv)
>   File "<C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\site-packages\decorator.py:decorator-gen-7>", line 2, in initialize
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\site-packages\traitlets\config\application.py", line 87, in catch_config_error
>     return method(app, *args, **kwargs)
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\site-packages\notebook\notebookapp.py", line 1679, in initialize
>     self.init_webapp()
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\site-packages\jupyterlab\labapp.py", line 404, in init_webapp
>     super().init_webapp(*args, **kwargs)
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\site-packages\notebook\notebookapp.py", line 1442, in init_webapp
>     self.http_server.listen(port, self.ip)
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\site-packages\tornado\tcpserver.py", line 152, in listen
>     self.add_sockets(sockets)
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\site-packages\tornado\tcpserver.py", line 165, in add_sockets
>     self._handlers[sock.fileno()] = add_accept_handler(
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\site-packages\tornado\netutil.py", line 279, in add_accept_handler
>     io_loop.add_handler(sock, accept_handler, IOLoop.READ)
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\site-packages\tornado\platform\asyncio.py", line 99, in add_handler
>     self.asyncio_loop.add_reader(fd, self._handle_events, fd, IOLoop.READ)
>   File "C:\WinP\bd38\bu\WPy64-3801b1\python-3.8.0.amd64\lib\asyncio\events.py", line 501, in add_reader
>     raise NotImplementedError
> NotImplementedError
> 
> C:\WinP\bd38\bu\WPy64-3801b1\notebooks>
> ```

I'm getting this too on Windows 10 with jupyterlab 1.2.3 and tornado 6.0.3 in a new conda-forge environment. I guess we need to wait for proper 3.8 support from tornado on Windows before we can use jupyterlab....

**Comment by mobbSF** [comment]
> maybe anybody can help solve this problem:
> ![image](https://user-images.githubusercontent.com/8581044/68591444-73f27b80-0499-11ea-80a3-7ba6e9ca9445.png)
> 
> so, why i can't use python3.8 at jupyter lab (v.1.2.2)?

looks like a walrus operator

**Comment by vidartf** [comment]
Closing, as JupyterLab is running fine on 3.8, and has been for a while. This issue is not about 3.8 kernels :)
